src/jit/codegenxarch.cpp

```diff
@@ -6841,47 +6887,142 @@ void        CodeGen::genCompareLong(GenTreePtr  treeNode)
     getEmitter()->emitInsBinary(ins, cmpAttr, hiOp1, hiOp2);
 
     // Generate the first jump for the high compare
-    emitJumpKind jumpKind[2];
-    bool branchToTrueLabel[2];
-    genJumpKindsForTreeLongHi(tree, jumpKind, branchToTrueLabel);
+    CompareKind compareKind = ((tree->gtFlags & GTF_UNSIGNED) != 0) ? CK_UNSIGNED : CK_SIGNED;
+    
 
-    BasicBlock* label = genCreateTempLabel();
-    inst_JMP(jumpKind[0], label);
+    BasicBlock* labelHi = genCreateTempLabel();
+    BasicBlock* labelFinal = genCreateTempLabel();
 
-    // Generate the second jump for LE, LT, GT, and GE. We only do the lower compare if
-    // the hi parts are equal
-    if (jumpKind[1] != EJ_NONE)
+    if (compareKind == CK_SIGNED && (tree->gtOper != GT_NE && tree->gtOper != GT_EQ))
     {
-        assert(branchToTrueLabel[1]);
-        inst_JMP(jumpKind[1], label);
-    }
+        if (targetReg == REG_NA)
+        {
+            // For uncontained compares (compares used by GT_JTRUE nodes), we need to generated all of the
+            // jump logic here. For GT_GT and GT_GE, we will generate:
+            //      cmp hi1, hi2
+            //      jg trueLabel
+            //      jl labelFinal
+            //      cmp lo1, lo2
+            //      ja(e) trueLabel
+            //      labelFinal:
+            // For GT_LT and GT_LE, we will generate:
+            //      cmp hi1, hi2
+            //      jl trueLabel
+            //      jg labelFinal
+            //      cmp lo1, lo2
+            //      jb(e) trueLabel
+            //      labelFinal:
 
```

Comment:
In these cases, there's an optimization we should add (you can add it as a TODO): note that labelFinal will be placed at the very end of a BasicBlock. There will be no code generated afterwards. If the next block is BBJ_ALWAYS, and has no code, we should set labelFinal = compiler->compCurBB->bbNext->bbJumpDest, and not generate a labelFinal label.